### DISCUSSION TITLE
ESP visualisation: ColourMaps and ContourAndDrawGrid

### DISCUSSION THREAD

**Discussion by Cajac102** [rdkit/rdkit#4542] [Q&A] [answered]
Hey,

I am currently trying to visualize electrostatic potentials of a molecule in a plane. 
So far, it looks like this:
![contour_from_py_3](https://user-images.githubusercontent.com/41333514/134161487-6011b2d9-33c9-447e-be09-97c8b6d85e07.png)

Code:
```
        tp = pd.read_csv("tp.xyz", skiprows = 11, names = ["x", "y", "z", "ESP"], delim_whitespace=True)
        x = np.linspace(-20, 20, 401)
        z = np.linspace(-20, 20, 401)

        esp_vals = tp["ESP"].to_numpy()
        esp_vals = np.reshape(esp_vals, (401,401))

        contour_level = 10
        L2= []
        for l in range(contour_level+1):
             L2.append(contour_level/200*(-1) + l*0.01)

        d = Draw.MolDraw2DSVG(300, 300)
        d.ClearDrawing()
        ps = Draw.ContourParams()
        ps.fillGrid = True
        ps.gridResolution = 0.01
        Draw.ContourAndDrawGrid(d, esp_vals, x.tolist(), z.tolist(), params=ps, levels=L2)
        d.drawOptions().clearBackground = False
        d.FinishDrawing()
        txt = d.GetDrawingText()
        with open("contour_from_py_3.svg", 'w+') as outf:
          print(txt, file=outf)

```

So far, so good. This is the look I aim for:
![Figure_5_small](https://user-images.githubusercontent.com/41333514/134162901-dac52256-7c37-4f8e-b48d-f4e7eec6fd2b.png)

Is this possible with rdkit? 
There is a `setColourMap`Function in the docs, but I don't understand what it expects as parameters and I did not find any examples. 

If it is not possible, do you have suggestions for other workarounds? 
My backup plan would be to use matplotlib (see second plot) and draw the molecule on top, but then I (probably?) can't use rdkits Draw functions. 

Cheers,
Caro

**Comment by greglandrum** [comment, marked as answer]
oh, that's fun. I'm not sure it's possible, but we can certainly try.
Can you upload a version of tp.xyz  (you'll probably have to change its name to tp.xyz.txt) so that I can try a few things?

**Reply by Cajac102** [reply]
[tp.xyz.txt](https://github.com/rdkit/rdkit/files/7202888/tp.xyz.txt)

I wasn't sure if github allows larger uploads (12MB), I hope it works :)

**Reply by greglandrum** [reply]
I think to get what you want you're going to need to use matplotlib directly and draw the molecule on top. There's a thread here which may be helpful: https://sourceforge.net/p/rdkit/mailman/rdkit-discuss/thread/424adc96-ca4c-3c6b-ebf3-edf5098f820b%40ucsd.edu/#msg36735785

FWIW, this was the best I could come up with... some of the limitations is that the contouring code takes either a 2-point or 3-point colormap (so it's not as flexible as what you are looking for), that the code is filling squares on the plot instead of the region between contours, and that you can't currently tell the code not to shade regions which are outside of a fixed max/min value (this is why the image is dark in the middle):
```
tp = pd.read_csv("~/Downloads/tp.xyz", skiprows = 11, names = ["x", "y", "z", "ESP"], delim_whitespace=True)
x = np.linspace(-20, 20, 401)
z = np.linspace(-20, 20, 401)


contour_level = 10
L2= []
for l in range(contour_level+1):
     L2.append(contour_level/200*(-1) + l*0.01)


esp_vals = tp["ESP"].to_numpy()
capped_vals = [L2[0] if x<L2[0] else L2[-1] if x> L2[-1] else x for x in esp_vals]        
           
capped_vals = np.reshape(capped_vals, (401,401))
    
d = Draw.MolDraw2DSVG(300, 300)
d.ClearDrawing()
ps = Draw.ContourParams()
ps.fillGrid = True
ps.gridResolution = 0.2
cmap = cm.PiYG
ps.setColourMap([tuple(x) for x in cmap([0,0.5,1])])
Draw.ContourAndDrawGrid(d, capped_vals, x.tolist(), z.tolist(), params=ps, levels=L2)
d.drawOptions().clearBackground = False
d.FinishDrawing()
txt = d.GetDrawingText()
SVG(txt)
```
The image is:
![image](https://user-images.githubusercontent.com/540511/134186100-1ff2e4a5-b7db-46d0-bad6-53be580a8b86.png)

**Reply by Cajac102** [reply]
Thanks for the quick help! 
I will have a look into the thread. Your example is also really nice to try out the ColourMap function.

Thanks again!
